Implement VSTS PR trigger support in dev/pipelinesupport branch of dotnet-ci

smile21prc opened the pull request:
Implement VSTS PR trigger support in dev/pipelinesupport branch of dotnet-ci

smile21prc commented:
@mmitche 
The changes in the plugin might not be necessary. As long as the run config has refspec generated correctly for the PR (as shown in the screenshot blow):
![withrefspeccorrectlygenerated](https://user-images.githubusercontent.com/17461693/30679423-55adfae0-9e4e-11e7-97c9-23bab204772a.jpg)

The run is executed successfully, like:
https://dotnet-vsts.westus2.cloudapp.azure.com/job/DevDiv_dotnet-linker/job/TestJobs/job/testing-pr-linux-Configuration_Release/15/console

mmitche commented:
Yep.  We just need to get the right refspec in the right places.

From: smile21prc [mailto:notifications@github.com]
Sent: Wednesday, September 20, 2017 9:55 PM
To: dotnet/dotnet-ci <dotnet-ci@noreply.github.com>
Cc: Matt Mitchell <mmitche@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [dotnet/dotnet-ci] Implement VSTS PR trigger support in dev/pipelinesupport branch of dotnet-ci (#819)
Importance: High


@mmitche<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmmitche&data=02%7C01%7Cmmitche%40microsoft.com%7Ccc8785720ddd40c9278d08d500acef5e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636415665115332256&sdata=KOmWmOH50V%2BZarSuR7OBtzDX38V%2BfH1ynaqV2fBwNjw%3D&reserved=0>
The changes in the plugin might not be necessary. As long as the run config has refspec generated correctly for the PR (as shown in the screenshot blow):
[withrefspeccorrectlygenerated]<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F17461693%2F30679423-55adfae0-9e4e-11e7-97c9-23bab204772a.jpg&data=02%7C01%7Cmmitche%40microsoft.com%7Ccc8785720ddd40c9278d08d500acef5e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636415665115332256&sdata=FkkI3FVUj0Ul1tOLfGjmOMXcoVcCbhNzAf9oKwsdyG8%3D&reserved=0>

The run is executed successfully, like:
https://dotnet-vsts.westus2.cloudapp.azure.com/job/DevDiv_dotnet-linker/job/TestJobs/job/testing-pr-linux-Configuration_Release/15/console<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdotnet-vsts.westus2.cloudapp.azure.com%2Fjob%2FDevDiv_dotnet-linker%2Fjob%2FTestJobs%2Fjob%2Ftesting-pr-linux-Configuration_Release%2F15%2Fconsole&data=02%7C01%7Cmmitche%40microsoft.com%7Ccc8785720ddd40c9278d08d500acef5e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636415665115332256&sdata=uBVt%2Fx3AjaboPrGq9AKRFWeOMMIINUqDXRkoAECNYuU%3D&reserved=0>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fdotnet-ci%2Fpull%2F819%23issuecomment-331050857&data=02%7C01%7Cmmitche%40microsoft.com%7Ccc8785720ddd40c9278d08d500acef5e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636415665115332256&sdata=0409XT0cHOg2Sjvh5q7%2Bzg54P6qvsjQPqbxlHsSIctg%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAIUissw20YaIW5sNGp56kWC_inZhCYB1ks5skewsgaJpZM4Pey48&data=02%7C01%7Cmmitche%40microsoft.com%7Ccc8785720ddd40c9278d08d500acef5e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636415665115332256&sdata=QtvVrS1mAxwacsCg7L4w8u0uxsINz6RkF2JSuM0se5U%3D&reserved=0>.

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -329,7 +329,7 @@ repos.each { repoInfo ->
                         if (isDSLPRFromSameRepo) {
                             if (isVSTS) {
-                                // TODO: VSTS PR refspec
+                                refspec('+refs/pull/${ghprbPullId}/merge:refs/remotes/pull/${ghprbPullId}/merge')
```
This refspec is wrong.  The variable ghprbPullId wouldn't be valid for VSTS here (it's set by the github pull request builder)

smile21prc replied:
As discussed, changed it to vstsRefspec, which is set by TFS Jenkins plugin depending on whether this is is a PR build.

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -348,7 +348,7 @@ repos.each { repoInfo ->
                     if (isDSLPRFromSameRepo) {
                         if (isVSTS) {
-                            // TODO: VSTS PR branch
+                            branch("*/${repoInfo.utilitiesRepoBranch}")
```
This needs to be the hash that is being PR'd

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -371,7 +371,7 @@ repos.each { repoInfo ->
                         if (isPRTest) {
                             if (isVSTS) {
-                                // TODO: VSTS PR refspec
+                                refspec('+refs/pull/${ghprbPullId}/merge:refs/remotes/pull/${ghprbPullId}/merge')
```
Same comment as above (re: ghprbPullId)

smile21prc replied:
As discussed, changed it to vstsRefspec, which is set by TFS Jenkins plugin depending on whether this is is a PR build.

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -394,7 +394,7 @@ repos.each { repoInfo ->
                     if (isPRTest) {
                         if (isVSTS) {
-                            // TODO: VSTS PR branch
+                            branch("*/${repoInfo.branch}")
```
This needs to be the hash that is being PR'd.

smile21prc replied:
Changed it to vstsBranch, which is set by TFS Jenkins plugin with the commit hash that is being PR'd.

smile21prc commented:
@mmitche changed as suggested. Would you please check whether they are as expected?

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -7,6 +7,7 @@ class VSTSPipelineScm implements PipelineScm {
     private String _branch
     private String _credentials
     private String _collection
+    boolean isVSTS = getScmType() == 'VSTS'
```
There is no need for this.  It will always be VSTS inside this file

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -394,7 +394,7 @@ repos.each { repoInfo ->
                     if (isPRTest) {
                         if (isVSTS) {
-                            // TODO: VSTS PR branch
+                            branch(vstsBranch)
```
This looks incorrect.  This is a variable here that will be expanded when this script runs, rather than when the generated job runs.  You want '${vstsBranch}' probably

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -24,7 +25,12 @@ class VSTSPipelineScm implements PipelineScm {
+            return Utilities.calculateVSTSGitURL(this._collection, this._project)
+        }
+        else {
+            return Utilities.calculateGitHubURL(this._project)
```
No need for this code.

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -37,11 +43,19 @@ class VSTSPipelineScm implements PipelineScm {
+                    stringParam('VSTSBranch', vstsBranch, 'VSTS commit hash')
+                }
+                else {
+                    stringParam('sha1', '', 'Incoming sha1 parameter from the GHPRB plugin.')
```
No need for this code.

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -59,10 +73,16 @@ class VSTSPipelineScm implements PipelineScm {
+                                    url('${VSTSRepoUrl}')
+                                }
+                                else {
+                                    refspec('${GitRefSpec}')
```
No need for this code.

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -329,7 +329,7 @@ repos.each { repoInfo ->
                         if (isDSLPRFromSameRepo) {
                             if (isVSTS) {
-                                // TODO: VSTS PR refspec
+                                refspec(${vstsRefspec})
```
You need to quote this:
`refspec('${vstsRefspec}')`

mmitche replied:
Are you also missing adding these parameters to the generator jobs?

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -348,7 +348,7 @@ repos.each { repoInfo ->
                     if (isDSLPRFromSameRepo) {
                         if (isVSTS) {
-                            // TODO: VSTS PR branch
+                            branch(${vstsBranchOrCommit})
```
You need to quote this:
`refspec('${vstsBranchOrCommit}')`

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -371,7 +371,7 @@ repos.each { repoInfo ->
                         if (isPRTest) {
                             if (isVSTS) {
-                                // TODO: VSTS PR refspec
+                                refspec(${vstsRefspec})
```
You need to quote this:
`refspec('${vstsRefspec}')`

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -394,7 +394,7 @@ repos.each { repoInfo ->
                     if (isPRTest) {
                         if (isVSTS) {
-                            // TODO: VSTS PR branch
+                            branch(${vstsBranchOrCommit})
```
You need to quote this:
`refspec('${vstsBranchOrCommit}')`

smile21prc commented:
Thanks @mmitche for the review, updated as suggested.

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -429,6 +429,8 @@ repos.each { repoInfo ->
                 if (isVSTS) {
                     stringParam('VSTSCollectionName', repoInfo.collection, 'VSTS collection name')
                     stringParam('VSTSCredentialsId', repoInfo.credentials, 'VSTS credentials id')
+                    stringParam('VSTSRefspec', vstsRefspec, 'VSTS refspec')
```
This is incorrect.  The code above is referencing the VSTSRefSpec parameter, that is defined here, but injected by your plugin.  The second parameter here should probably be empty (default refspec).

Also, parameters are case sensitive, so be sure to match the name of this parameter with what your plugin is injecting

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -429,6 +429,8 @@ repos.each { repoInfo ->
                     stringParam('VSTSCollectionName', repoInfo.collection, 'VSTS collection name')
                     stringParam('VSTSCredentialsId', repoInfo.credentials, 'VSTS credentials id')
+                    stringParam('VSTSRefspec', vstsRefspec, 'VSTS refspec')
+                    stringParam('VSTSRBranchOrCommit', vstsBranchOrCommit, 'VSTS branch or commit')
```
See above

mmitche commented on src/jobs/generation/Utilities.groovy:
```diff
@@ -616,6 +616,20 @@ class Utilities {
+    def static addVSTSPRTrigger(def job) {
+        job.with {
+            triggers {
+                TeamPRPushTrigger()
```
Is the casing correct here?  Why is it camelcase for teamPushTrigger but TeamPRPushTrigger is different?

mmitche commented on src/jobs/generation/Utilities.groovy:
```diff
@@ -616,6 +616,20 @@ class Utilities {
     /**
+     * Adds a VSTS PR trigger
+     */
+    def static addVSTSPRTrigger(def job) {
```
Shouldn't there be a context passed in here for display in the VSTS UI?

mmitche commented on src/org/dotnet/ci/pipelines/Pipeline.groovy:
```diff
@@ -276,6 +276,18 @@ class Pipeline {
+    //  parameter - Optional set of key/value pairs of string parameters that will be passed to the pipeline
+    public def triggerPipelineOnEveryVSTSPR(String context, Map<String,Object> parameters = [:]) {
+        // Create a trigger builder and pass it to the generic triggerPipelineOnEvent
+        VSTSTriggerBuilder builder = VSTSTriggerBuilder.triggerOnPullRequest()
```
Shouldn't the context be passed in here?

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -37,11 +37,11 @@ class VSTSPipelineScm implements PipelineScm {
+                stringParam('VSTSCredentialsId', repoInfo.credentials, 'VSTS credentials id')
+                stringParam('VSTSRepoUrl', this.getGitUrl(), 'VSTS repo to clone.')
+                stringParam('VSTSRefspec', vstsRefspec, 'VSTS refspec')
+                stringParam('VSTSBranchOrCommit', vstsBranchOrCommit, 'VSTS commit hash')
```
Does the casing of VSTSRefSpec and VSTSBranchOrCommit match the plugin's casing?

mmitche commented on src/org/dotnet/ci/triggers/VSTSTriggerBuilder.groovy:
```diff
@@ -74,22 +74,28 @@ class VSTSTriggerBuilder implements TriggerBuilder {
+    def private emitPRTrigger(def job) {
+        job.with {
+            triggers {
+                teamPRPushTrigger()
```
This should probably pass the context in.

smile21prc commented:
Thanks @mmitche for further reviews. I've addressed all your comments.

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -429,6 +429,8 @@ repos.each { repoInfo ->
                     stringParam('VSTSCollectionName', repoInfo.collection, 'VSTS collection name')
                     stringParam('VSTSCredentialsId', repoInfo.credentials, 'VSTS credentials id')
+                    stringParam('vstsRefspec', '', 'VSTS refspec')
+                    stringParam('vstsBranchOrCommit', '*/master', 'VSTS branch or commit')
```
Instead of master, it should be the configured branch for this entry in the MetaGenerator: repoInfo.branch

mmitche commented on src/jobs/generation/MetaGenerator.groovy:
```diff
@@ -508,7 +510,7 @@ repos.each { repoInfo ->
         if (isPRTest) {
             if (isVSTS) {
-                // TODO: VSTS PR trigger
+                Utilities.addVSTSPRTrigger(jobGenerator, "Gen CI(${repoInfo.server}) - ${repoInfo.branch}/${repoInfo.definitionScript}")
```
Will then end up being branch specific?  As in, if I make a PR to release/2.0.0 to change the CI definition, is it going to run a generator intended for master?

smile21prc replied:
Changed the groovy script to be branch specific for PR. Will add branch support for TFS plugin as well.

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -37,13 +37,13 @@ class VSTSPipelineScm implements PipelineScm {
+                stringParam('VSTSCredentialsId', this._credentials, 'VSTS credentials id')
+                stringParam('VSTSRepoUrl', this.getGitUrl(), 'VSTS repo to clone.')
+                stringParam('vstsRefspec', '', 'VSTS refspec')
+                stringParam('vstsBranchOrCommit', '*/master', 'VSTS commit hash')
```
Instead of master, it should be the configured branch.  See the Github SCM for an example

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -60,9 +60,9 @@ class VSTSPipelineScm implements PipelineScm {
                                 url(this.getGitUrl())
                                 // Set the refspec to be the parmeterized version
-                                refspec('${GitRefSpec}')
+                                refspec('${VSTSRefspec}')
```
This still looks like incorrect casing.

smile21prc replied:
Good catch, these are left-over.

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -60,9 +60,9 @@ class VSTSPipelineScm implements PipelineScm {
+                                refspec('${VSTSRefspec}')
                                 // Set URL to the parameterized version
-                                url('${GitRepoUrl}')
+                                url('${VSTSRepoUrl}')
```
Still incorrect casing I think.

smile21prc replied:
VSTSRepoUrl matches the casing above:                 
stringParam('VSTSRepoUrl', this.getGitUrl(), 'VSTS repo to clone.')

Which should be necessary as GithubPipelineScm.groovy has similar one:
stringParam('GitRepoUrl', Utilities.calculateGitURL(this._project), 'Git repo to clone.')

smile21prc commented:
Thanks @mmitche for additional comments. Added support for branch-specific triggering and addressed your other comments as well. Could you please take another look?

mmitche commented:
@dotnet-bot test this please (testing something)

smile21prc commented:
@dotnet-bot test this please

smile21prc commented:
@dotnet-bot test this please (testing something)

smile21prc commented:
@dotnet-bot test this please (testing something)

mmitche commented on src/org/dotnet/ci/pipelines/scm/VSTSPipelineScm.groovy:
```diff
@@ -37,13 +37,13 @@ class VSTSPipelineScm implements PipelineScm {
+                stringParam('VSTSCredentialsId', this._credentials, 'VSTS credentials id')
+                stringParam('VSTSRepoUrl', this.getGitUrl(), 'VSTS repo to clone.')
+                stringParam('vstsRefspec', '', 'VSTS refspec')
+                stringParam('vstsBranchOrCommit', '*/${this._branch}', 'VSTS commit hash')
```
This needs to be double quoted to get string interpolation

smile21prc commented:
Good catch @mmitche . Fixed it and will merged it.
